Crash on App Open on Xiaomi A1

app version that crashes: 0.8.0
Crash occurs after PIN lock enter

Crash log:

```
java.lang.RuntimeException: 
  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run (RuntimeInit.java:503)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:858)
Caused by: java.lang.reflect.InvocationTargetException: 
  at java.lang.reflect.Method.invoke (Native Method)
  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run (RuntimeInit.java:493)
Caused by: javax.crypto.BadPaddingException: 
  at android.security.keystore.AndroidKeyStoreCipherSpiBase.engineDoFinal (AndroidKeyStoreCipherSpiBase.java:515)
  at javax.crypto.Cipher.doFinal (Cipher.java:2055)
  at io.horizontalsystems.bankwallet.core.security.CipherWrapper.decrypt (CipherWrapper.kt:40)
  at io.horizontalsystems.bankwallet.core.security.EncryptionManager.decrypt (EncryptionManager.kt:17)
  at io.horizontalsystems.bankwallet.core.managers.SecuredStorageManager.getSavedPin (SecuredStorageManager.kt:29)
  at io.horizontalsystems.bankwallet.core.managers.PinManager.validate (PinManager.kt:18)
  at io.horizontalsystems.bankwallet.modules.pin.unlock.UnlockPinInteractor.unlock (UnlockPinInteractor.kt:33)
  at io.horizontalsystems.bankwallet.modules.pin.unlock.UnlockPinPresenter.onEnter (UnlockPinPresenter.kt:37)
  at io.horizontalsystems.bankwallet.modules.pin.PinActivity.onItemClick (PinActivity.kt:177)
  at io.horizontalsystems.bankwallet.ui.extensions.NumPadItemsAdapter$onBindViewHolder$1.invoke (NumPadItemsAdapter.kt:48)
  at io.horizontalsystems.bankwallet.ui.extensions.NumPadItemsAdapter$onBindViewHolder$1.invoke (NumPadItemsAdapter.kt:13)
  at io.horizontalsystems.bankwallet.ui.extensions.NumPadItemViewHolder$bind$1.onTouch (NumPadItemsAdapter.kt:74)
```

--- a/app/src/main/java/io/horizontalsystems/bankwallet/core/Interfaces.kt
+++ b/app/src/main/java/io/horizontalsystems/bankwallet/core/Interfaces.kt
@@ -160,8 +160,6 @@ interface IKeyStoreManager {
     val isKeyInvalidated: Boolean
     val isUserNotAuthenticated: Boolean
 
-    fun createKey(): SecretKey
-    fun getKey(): SecretKey
     fun removeKey()
 }
 

--- a/app/src/main/java/io/horizontalsystems/bankwallet/core/managers/AccountManager.kt
+++ b/app/src/main/java/io/horizontalsystems/bankwallet/core/managers/AccountManager.kt
@@ -25,6 +25,7 @@ class AccountManager(private val storage: IAccountsStorage, private val accountC
 
     override val accountsFlowable: Flowable<List<Account>>
         get() = accountsSubject.toFlowable(BackpressureStrategy.BUFFER)
+
     override val deleteAccountObservable: Flowable<String>
         get() = deleteAccountSubject.toFlowable(BackpressureStrategy.BUFFER)
 
@@ -59,10 +60,11 @@ class AccountManager(private val storage: IAccountsStorage, private val accountC
     }
 
     override fun clear() {
-        storage.clear()
+        cache.accountsSet.map { it.id }.forEach { accountId ->
+            storage.delete(accountId)
+        }
         cache.set(listOf())
-
-        accountsSubject.onNext(accounts)
+        accountsSubject.onNext(listOf())
     }
 
     override fun clearAccounts() {

--- a/app/src/main/java/io/horizontalsystems/bankwallet/core/managers/KeyStoreChangeListener.kt
+++ b/app/src/main/java/io/horizontalsystems/bankwallet/core/managers/KeyStoreChangeListener.kt
@@ -22,6 +22,4 @@ class KeyStoreChangeListener(private val systemInfoManager: ISystemInfoManager,
         }
     }
 
-    override fun didEnterBackground() {}
-
 }

--- a/app/src/main/java/io/horizontalsystems/bankwallet/core/security/KeyStoreManager.kt
+++ b/app/src/main/java/io/horizontalsystems/bankwallet/core/security/KeyStoreManager.kt
@@ -47,21 +47,6 @@ class KeyStoreManager(private val keyAlias: String) : IKeyStoreManager, IKeyProv
             ex is UserNotAuthenticatedException
         }
 
-    override fun createKey(): SecretKey {
-        val keyGenerator = KeyGenerator.getInstance(KeyProperties.KEY_ALGORITHM_AES, ANDROID_KEY_STORE)
-
-        keyGenerator.init(KeyGenParameterSpec.Builder(keyAlias,
-                KeyProperties.PURPOSE_ENCRYPT or KeyProperties.PURPOSE_DECRYPT)
-                .setBlockModes(BLOCK_MODE)
-                .setUserAuthenticationRequired(true)
-                .setUserAuthenticationValidityDurationSeconds(AUTH_DURATION_SEC)
-                .setRandomizedEncryptionRequired(false)
-                .setEncryptionPaddings(PADDING)
-                .build())
-
-        return keyGenerator.generateKey()
-    }
-
     override fun getKey(): SecretKey {
         val key = keyStore.getKey(keyAlias, null) ?: createKey()
         return key as SecretKey
@@ -73,6 +58,20 @@ class KeyStoreManager(private val keyAlias: String) : IKeyStoreManager, IKeyProv
         } catch (ex: KeyStoreException) {
             logger.warning("removeKey: \n ${ex.getStackTraceString()}")
         }
+    }
+
+    private fun createKey(): SecretKey {
+        val keyGenerator = KeyGenerator.getInstance(KeyProperties.KEY_ALGORITHM_AES, ANDROID_KEY_STORE)
+
+        keyGenerator.init(KeyGenParameterSpec.Builder(keyAlias, KeyProperties.PURPOSE_ENCRYPT or KeyProperties.PURPOSE_DECRYPT)
+                .setBlockModes(BLOCK_MODE)
+                .setUserAuthenticationRequired(true)
+                .setUserAuthenticationValidityDurationSeconds(AUTH_DURATION_SEC)
+                .setRandomizedEncryptionRequired(false)
+                .setEncryptionPaddings(PADDING)
+                .build())
+
+        return keyGenerator.generateKey()
     }
 
     private fun validateKey() {

--- a/app/src/main/java/io/horizontalsystems/bankwallet/modules/keystore/KeyStoreInteractor.kt
+++ b/app/src/main/java/io/horizontalsystems/bankwallet/modules/keystore/KeyStoreInteractor.kt
@@ -23,8 +23,6 @@ class KeyStoreInteractor(private val accountManager: IAccountManager,
         accountManager.clear()
         walletManager.enable(listOf())
         localStorage.clear()
-
-        //TODO clear kits
     }
 
     override fun removeKey() {

--- a/app/src/main/java/io/horizontalsystems/bankwallet/modules/keystore/KeyStorePresenter.kt
+++ b/app/src/main/java/io/horizontalsystems/bankwallet/modules/keystore/KeyStorePresenter.kt
@@ -1,29 +1,33 @@
 package io.horizontalsystems.bankwallet.modules.keystore
 
+import io.horizontalsystems.bankwallet.modules.keystore.KeyStoreModule.ModeType
+
 class KeyStorePresenter(private val interactor: KeyStoreModule.IInteractor,
                         private val router: KeyStoreModule.IRouter,
-                        private val mode: KeyStoreModule.ModeType) : KeyStoreModule.IViewDelegate, KeyStoreModule.IInteractorDelegate {
+                        private val mode: ModeType) : KeyStoreModule.IViewDelegate, KeyStoreModule.IInteractorDelegate {
 
     var view: KeyStoreModule.IView? = null
 
     override fun viewDidLoad() {
         when (mode) {
-            KeyStoreModule.ModeType.NoSystemLock -> {
+            ModeType.NoSystemLock -> {
                 interactor.resetApp()
                 view?.showNoSystemLockWarning()
             }
-            KeyStoreModule.ModeType.InvalidKey -> {
+            ModeType.InvalidKey -> {
                 interactor.resetApp()
                 view?.showInvalidKeyWarning()
             }
-            KeyStoreModule.ModeType.UserAuthentication -> view?.promptUserAuthentication()
+            ModeType.UserAuthentication -> {
+                view?.promptUserAuthentication()
+            }
         }
     }
 
     override fun onResume() {
-        if (!interactor.isSystemLockOff &&
-                !interactor.isKeyInvalidated &&
-                !interactor.isUserNotAuthenticated) {
+        if (mode != ModeType.UserAuthentication &&
+                !interactor.isSystemLockOff &&
+                !interactor.isKeyInvalidated) {
             interactor.removeKey()
             router.openLaunchModule()
         }

